R/pcrelate.R

```diff
@@ -128,18 +125,30 @@ setMethod("pcrelate",
         nsnpblock <- length(snp.blocks)
         if(verbose) message('Running PC-Relate analysis for ', length(sample.include), ' samples using ', length(unlist(snp.blocks)), ' SNPs in ', nsnpblock, ' blocks...')
 
+        # iterator function to return genotype blocks
+        k <- 1
+        Giter <- function() {
+            if (k > nsnpblock) return(NULL)
+            G <- .readGeno(gdsobj, sample.include, snp.index = snp.blocks[[k]])
+            k <<- k + 1
+            return(G)
+        }
+        
         # for each snp block
-        matList <- foreach(k = 1:nsnpblock, .combine = .matListCombine, .inorder = FALSE, .multicombine = FALSE) %do% {
-            if(verbose) message('    Running block ', k, '...')
+        #matList <- foreach(k = 1:nsnpblock, .combine = .matListCombine, .inorder = FALSE, .multicombine = FALSE) %do% {
+       snpBlock <- function(G, ...) {
+            #if(verbose) message('    Running block ', k, '...')
             # read genotype data for the block
-            G <- .readGeno(gdsobj, sample.include, snp.index = snp.blocks[[k]])
+            #G <- .readGeno(gdsobj, sample.include, snp.index = snp.blocks[[k]])
 
             # calculate ISAF betas
             beta <- .calcISAFBeta(G = G, VVtVi = VVtVi)
 
             # calculate PC-Relate estimates
             .pcrelateVarBlock(G = G, beta = beta, V = V, idx = 1:nrow(V), jdx = 1:nrow(V), scale = scale, ibd.probs = ibd.probs, maf.thresh = maf.thresh, maf.bound.method = maf.bound.method)
         }
+        
+        matList <- bpiterate(Giter, snpBlock, REDUCE = .matListCombine, reduce.in.order = FALSE, BPPARAM = BPPARAM)
 
```

Comment:
I think I probably just don't understand how `bpiterate` works, but why don't `VVtVi`, `V`, etc. need to be passed into the function here? I think the extra arguments were passed into `bpiterate` in the assoc functions.

Reply:
In the assoc functions, the function called by `bpiterate` is defined outside of the `assocTest*` function, so it can be used by multiple methods. Here, it's defined inside the `pcrelate` function, so all those variables are already in scope.